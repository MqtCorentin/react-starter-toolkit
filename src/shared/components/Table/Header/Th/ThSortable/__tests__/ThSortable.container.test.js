import React from 'react';
import { render } from '@testing-library/react';
import ThSortableContainer, { toggleOrder, toggleSorting, setOrder, SortingIcon, setClassModifierActive } from '../ThSortable.container';

describe('setClassModifierActive', () => {
  it('Should return "classModifier" when isActive equal false', () => {
    const result = setClassModifierActive({ classModifier: 'classModifier', isActive: false });
    expect(result).toEqual('classModifier');
  });
  it('Should return "classModifier active" when isActive equal true', () => {
    const result = setClassModifierActive({ classModifier: 'classModifier', isActive: true });
    expect(result).toEqual('classModifier active');
  });
});

describe('SortingIcon', () => {
  it('Should render SortingIcon with order equal NONE', () => {
    const { asFragment } = render(<SortingIcon order="NONE" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should render SortingIcon with order equal true', () => {
    const { asFragment } = render(<SortingIcon order />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('Should render SortingIcon with order equal false', () => {
    const { asFragment } = render(<SortingIcon order={false} />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('toggleOrder', () => {
  it('Should return 1 when receive -1', () => {
    const result = toggleOrder(-1);
    expect(result).toEqual(1);
  });

  it('Should return -1 when receive 1', () => {
    const result = toggleOrder(1);
    expect(result).toEqual(-1);
  });
});

describe('setOrder', () => {
  it.each`
    field       | prevField   | order | expected
    ${'champ1'} | ${'champ2'} | ${1}  | ${1}
    ${'champ1'} | ${'champ2'} | ${-1} | ${1}
    ${'champ1'} | ${'champ1'} | ${1}  | ${-1}
    ${'champ1'} | ${'champ1'} | ${-1} | ${1}
    ${null}     | ${'champ1'} | ${-1} | ${1}
    ${null}     | ${null}     | ${-1} | ${1}
    ${'champ1'} | ${null}     | ${-1} | ${1}
  `('Should return $expected when field: $field, prevField: $prevField, order: $order', ({ field, prevField, order, expected }) => {
    const result = setOrder({ field, prevField, order });
    expect(result).toEqual(expected);
  });
});

describe('toggleSorting', () => {
  it.each`
    field       | sorting                           | expected
    ${'champ1'} | ${{ field: 'champ2', order: -1 }} | ${{ field: 'champ1', order: 1 }}
    ${'champ1'} | ${{ field: 'champ2', order: 1 }}  | ${{ field: 'champ1', order: 1 }}
    ${'champ2'} | ${{ field: 'champ2', order: 1 }}  | ${{ field: 'champ2', order: -1 }}
    ${null}     | ${{ field: 'champ2', order: 1 }}  | ${{ field: null, order: 1 }}
    ${'champ1'} | ${{ field: null, order: 1 }}      | ${{ field: 'champ1', order: 1 }}
    ${'champ1'} | ${{ field: null, order: null }}   | ${{ field: 'champ1', order: 1 }}
  `('Should return $expected when field: $field, sorting: $sorting', ({ field, sorting, expected }) => {
    const result = toggleSorting({ field, sorting });
    expect(result).toEqual(expected);
  });
});

const containerTr = document.createElement('tr');

const renderWithContainer = Component =>
  render(<Component />, {
    container: document.body.appendChild(containerTr),
  });

const onSortMock = jest.fn().mockReturnValue('ordering');
const defaultProps = { children: 'sortable', sortable: false, onSort: onSortMock, field: '', sorting: { field: '', order: 'NONE' } };

describe('<ThSortableContainer />', () => {
  it('Should render ThSortableContainer with no sortable button', () => {
    const { asFragment, getByRole } = renderWithContainer(() => <ThSortableContainer {...defaultProps} />);
    expect(getByRole('button')).toBeDefined();
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should render ThSortableContainer with sortable button for firstname column', () => {
    const { asFragment, container } = renderWithContainer(() => (
      <ThSortableContainer {...defaultProps} sortable sorting={{ field: 'name', order: true }} field="name">
        sortable
      </ThSortableContainer>
    ));
    expect(container.querySelector('.glyphicon-arrow-xs-up')).toBeDefined();
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should call onSort function', () => {
    const toggleSortingMock = jest.fn();
    const setSortingIconMock = jest.fn();
    const setClassModifierActiveMock = jest.fn().mockReturnValue('classnameactive');

    const defaultPropsOverride = {
      ...defaultProps,
      toggleSortingFn: toggleSortingMock,
      setSortingIconFn: setSortingIconMock,
      setClassModifierActiveFn: setClassModifierActiveMock,
      sortable: true,
    };

    renderWithContainer(() => <ThSortableContainer {...defaultPropsOverride} />);

    expect(ThSortableContainer({ ...defaultPropsOverride }).props.sort()).toEqual('ordering');

    expect(toggleSortingMock).toBeCalled();
  });
});
