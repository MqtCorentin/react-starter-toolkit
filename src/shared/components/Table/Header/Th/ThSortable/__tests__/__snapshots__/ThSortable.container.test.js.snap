// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ThSortableContainer /> Should render ThSortableContainer with no sortable button 1`] = `
<DocumentFragment>
  sortable
  <span
    class="af-btn__icon af-btn__icon--table-sorting glyphicon glyphicon-sorting"
  />
</DocumentFragment>
`;

exports[`<ThSortableContainer /> Should render ThSortableContainer with sortable button for firstname column 1`] = `
<DocumentFragment>
  sortable
  <span
    class="af-btn__icon af-btn__icon--table-sorting glyphicon glyphicon-undefined"
  />
</DocumentFragment>
`;

exports[`SortingIcon Should render SortingIcon with order equal NONE 1`] = `
<DocumentFragment>
  <span
    class="af-btn__icon af-btn__icon--table-sorting glyphicon glyphicon-sorting"
  />
</DocumentFragment>
`;

exports[`SortingIcon Should render SortingIcon with order equal false 1`] = `
<DocumentFragment>
  <span
    class="af-btn__icon af-btn__icon--table-sorting glyphicon glyphicon-undefined"
  />
</DocumentFragment>
`;

exports[`SortingIcon Should render SortingIcon with order equal true 1`] = `
<DocumentFragment>
  <span
    class="af-btn__icon af-btn__icon--table-sorting glyphicon glyphicon-undefined"
  />
</DocumentFragment>
`;
